opentracing/src/main/java/opentracing/SpanPropagator.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+import java.util.Map;
+
+/**
+ * Encodes or Decodes a {@linkplain Span} in binary or text formats.
+ *
+ * <p>The toXXX methods are expected to serialize spans into a pair of values representing
+ * separately the span / span identity, and the trace attributes. This is done specifically
+ * for binary protocols that may represent tracing identity in a dedicated fixed-length slot in the
+ * binary message format, so that it can be inspected efficiently by the middleware / routing layers
+ * without parsing the whole message.
+ */
+public interface SpanPropagator {
+
+  /**
+   * Encodes the span into a binary representation of the span's identity and trace
+   * attributes.
+   */
+  byte[] toBinary(Span span);
```

Comment:
> This is the crux of the problem. You don't think of RPC framework authors as users of OpenTracing, yet they are.

Yes this has been an on-going misunderstanding, and i've tried a number of times to bring about better nomenclature into the discussion to help progress it. 

1. The end-user is the application developer. The api they use is the public API/SPI.
2. The authors of OT implementations. The api they use is the internal API, and do so to flesh out the public SPIs.

End-users read the Specification to gain an understanding on how to use the public API.
Authors read the Specification to know what their implementation must adhere to.

The java interfaces being introduced here are proposed as only the public API.

The development, evolution, and stability of the public and internal APIs is different, so not only will using clear nomenclature between the two help the discussions, but help us not shoot ourselves in the foot with API compatibilities issues over time. And lastly, by separating them we focus on building the simplest and most attractive end-user API possible.

I am definitely not saying that the author does not matter, but rather giving them their own arena.